python-package/securexgboost/remote_server.py

```diff
@@ -64,7 +64,300 @@ def is_ready(self):
+                        booster_handle=booster_handle,
+                        flen=flen,
+                        fname=fname,
+                        ftype=ftype,
+                        with_stats=with_stats,
+                        dump_format=dump_format
+                        ))
+                elif self._func == remote_api.XGBoosterGetModelRaw:
+                    booster_handle = self._params.booster_handle
+                    username = self._params.username
+                    response_future = stub.rpc_XGBoosterGetModelRaw.future(remote_pb2.ModelRawParams(
+                        booster_handle=booster_handle,
+                        username=username
+                        ))
+                elif self._func == remote_api.XGDMatrixNumRow:
+                    name = self._params.name
+                    response_future = stub.rpc_XGDMatrixNumRow.future(remote_pb2.Name(
+                        name=name
+                        ))
+                elif self._func == remote_api.XGDMatrixNumCol:
+                    name = self._params.name
+                    response_future = stub.rpc_XGDMatrixNumCol.future(remote_pb2.Name(
+                        name=name
+                        ))
+                elif self._func == remote_api.XGBoosterPredict:
+                    booster_handle = self._params.booster_handle
+                    dmatrix_handle = self._params.dmatrix_handle
+                    option_mask = self._params.option_mask
+                    ntree_limit = self._params.ntree_limit
+                    username = self._params.username
+                    response_future = stub.rpc_XGBoosterPredict.future(remote_pb2.PredictParams(
+                        booster_handle=booster_handle,
+                        dmatrix_handle=dmatrix_handle,
+                        option_mask=option_mask,
+                        ntree_limit=ntree_limit,
+                        username=username
+                        ))
+                futures.append(response_future)
+        
+            results = []
+            for future in futures:
+                results.append(future.result())
+
+            # Set return value
+            if self._func == rabit_remote_api.RabitInit:
+                return_codes = [result.status for result in results]
+                if sum(return_codes) == 0:
+                    self._ret = 0
+                else:
+                    self._ret = -1
+            elif self._func == rabit_remote_api.RabitFinalize:
+                return_codes = [result.status for result in results]
+                if sum(return_codes) == 0:
+                    self._ret = 0
+                else:
+                    self._ret = -1
+            elif self._func == remote_api.XGDMatrixCreateFromEncryptedFile:
+                statuses = [result.status.status for result in results]
+                if -1 in statuses:
```

Comment:
Perhaps consolidate the error checking code into a single location? All the RPC calls now return a `status`, so perhaps this check should be done once outside the `if...else` ladder, and not checked per function.

Reply:
I thought about doing this, but I don't think it's possible because some functions returns a different number of items than other functions. For example, XGBoosterCreate returns a booster handle and a status, while GetModelRaw returns `lengths`, `sarrs`, and a status

Reply:
Not sure I follow... The error status doesn't depend on the number of returned items, right? All it does is check if any `status` in the results is -1.  So if we detects a negative status, we can return. Otherwise, do the `if...else`.

Reply:
We need to set `self._ret` properly according to the function that is called. `self._ret` may contain a different number of items depending on the invoked API, and needs to be set accordingly if we detect a negative status -- that's why I did this error checking on a per function basis instead of once before this block.